### PR TITLE
Add udp packet

### PR DESCRIPTION
**Pull request by kyrofa**

### CODE REVIEW

**Comment by kyrofa** [comment]
Don't merge until #2 has been accepted and this has been rebased.

**Comment by kyrofa** [comment]
Alright this one is ready for review.

File: src/udp_packet.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+UdpPacket::UdpPacket(const std::vector<uint8_t> &buffer)
+{
+	auto bufferSize = buffer.size();
+	if (bufferSize < UDP_HEADER_SIZE)
+	{
+		std::stringstream stream;
+		stream << "UDP header is " << UDP_HEADER_SIZE << " bytes, but buffer "
+			   << "is only " << bufferSize << " bytes";
```
Comment (cakekoa) [inline src/udp_packet.cpp]: I don't believe you need the stream operator at the beginning of this line. Also, it's indented strangely.

File: include/udp_packet.h [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+		/*!
+		 * \brief Convert this packet into bytes.
+		 *
+		 * \param[out] buffer
+		 * The buffer which will be filled with bytes repesenting this packet.
+		 * Note that the bytes will be in network-byte-order.
+		 */
+		void toBuffer(std::vector<uint8_t> &buffer) const;
+
+		void toString(std::string &packetString) const;
```
Comment (cakekoa) [inline include/udp_packet.h:62]: Want to comment this function as well?

File: src/udp_packet.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	// Parse destination port, bytes 2-3.
+	m_destinationPort = (buffer[2] << 8) | buffer[3];
+
+	// Parse packet length, bytes 4-5.
+	m_packetLength = (buffer[4] << 8) | buffer[5];
+	if (bufferSize != m_packetLength)
+	{
+		std::stringstream stream;
+		stream << "Buffer is " << bufferSize << " bytes, expected "
+			   << m_packetLength;
```
Comment (cakekoa) [inline src/udp_packet.cpp]: This line also has some strange indentation.

File: src/udp_packet.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+#include "udp_packet.h"
+
+namespace
+{
+	const int UDP_HEADER_SIZE = 8;
+}
+
+UdpPacket::UdpPacket(uint16_t sourcePort, uint16_t destinationPort,
+					 const std::vector<uint8_t> &payload) :
```
Comment (cakekoa) [inline src/udp_packet.cpp]: Indentation

File: include/udp_packet.h [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+		 * The port from which this packet is being sent.
+		 *
+		 * \param[in] destinationPort
+		 * The port to which this packet is being sent.
+		 *
+		 * \param[in payload
+		 * The payload contained within this packet.
+		 */
+		UdpPacket(uint16_t sourcePort, uint16_t destinationPort,
+				  const std::vector<uint8_t> &payload);
```
Comment (cakekoa) [inline include/udp_packet.h]: Indentation

File: include/udp_packet.h [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+		/*!
+		 * \brief Construct a UdpPacket given its components.
+		 *
+		 * \param[in] sourcePort
+		 * The port from which this packet is being sent.
+		 *
+		 * \param[in] destinationPort
+		 * The port to which this packet is being sent.
+		 *
+		 * \param[in payload
```
Comment (cakekoa) [inline include/udp_packet.h]: [in]

File: src/udp_packet.cpp (hunk 1 above)
Comment (kyrofa) [inline src/udp_packet.cpp]: Yeah my indentation is all over the place due to [this age-old bug](https://bugreports.qt.io/browse/QTCREATORBUG-7994). Should be better now.

As for the stream operator, I just feel like it lends itself to better readability (when the indentation is correct).

**Comment by kyrofa** [comment]
Alright I believe I addressed the issues you noticed.

**Pull request merged**